### ISSUE TITLE
[FEA] Agg groupby.filter support

### ISSUE THREAD

**Issue by miguelusque** [rapidsai/cudf#7354]
**Is your feature request related to a problem? Please describe.**
Hi!

While porting some existing code from Pandas to cuDF, I have just noticed that `groupby.filter` method is not supported.

The code I am porting is the following:

> import pandas as pd
> df1 = pd.read_csv('./data/ml-20m/ratings.csv')
> df1 = df1.groupby(USER_COLUMN).filter(lambda x: len(x) >= MIN_RATINGS)

I have already found a workaround for it, but I think it would be great to add support to `groupby.filter` method.

**Describe the solution you'd like**
Having `groupby.filter` method available in cudf.

Thanks!

Miguel

**Comment by kkraus14** [comment]
Relevant Pandas API Docs: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.filter.html

**Comment by shwina** [comment]
Similar to `apply()` the best approach we have right now for this is to iterate over the groups sequentially and apply the UDF on each group. This will only be efficient on the GPU for a small number of large groups (and slow for a large number of large groups).

**Comment by harrism** [comment]
>  This will only be efficient on the GPU for a small number of large groups (and slow for a large number of small groups).

But can be improved with multi-stream concurrency, which we plan to add in the future.

If we can use Numba to compile a (limited) filter syntax to PTX, we could apply this in parallel in a single kernel. How general are the filters that people want to run? Is it as broad as `apply()`?

**Comment by kkraus14** [comment]
> How general are the filters that people want to run? Is it as broad as `apply()`?

Yes, it's unfortunately similar to apply where it expects a function that takes in a DataFrame and returns either a scalar boolean value or a column of boolean values. Because the function expects to take in a DataFrame, it can use completely arbitrary DataFrame APIs so there's no easy way for us to pipeline it through Numba.

**Comment by shwina** [comment]
Note that for many simple cases, there's a way to work around the lack of support for `groupby.filter` using `groupby.apply` + `merge`:

```python
def groupby_filter(df, key, func):
    if not isinstance(key, list):
        key = [key]
    keep_group = df.groupby(key).apply(func).to_frame()
    keep_group.columns = ["_"]
    keep_row = df[key].merge(keep_group, left_on=key, right_index=True)["_"].sort_index()
    return df.loc[keep_row]
```

Example usage:

```python
>>> df = cudf.datasets.randomdata(100_000)
>>> pdf = df.to_pandas()
```

```python
# Pandas:
>>> pdf.groupby('id').filter(lambda df: (df.x > df.y).all())                                                                                                                                              
         id         x         y
5222   1106 -0.394735 -0.787261
16999  1120  0.849243  0.171776
17870  1113 -0.270606 -0.938197
20487  1106  0.392992 -0.868924
24426   863 -0.199523 -0.831697
33665  1124  0.915199 -0.033524
34860  1122  0.305254  0.244650
37147  1122  0.982420 -0.252851
53573  1119 -0.457173 -0.575891
63077   874  0.914197 -0.133111
72263  1106  0.924446  0.187110
75129  1126 -0.820924 -0.831940
77157   874  0.286803 -0.908891
80028  1117  0.182344 -0.297617
85099  1124  0.454558  0.011887
93097   889  0.333730 -0.539914
93482  1106 -0.448811 -0.646975
```

```python
# cuDF:
In [39]: groupby_filter(df, 'id', lambda group: (group.x > group.y).all())
Out[39]: 
         id         x         y
5222   1106 -0.394735 -0.787261
16999  1120  0.849243  0.171776
17870  1113 -0.270606 -0.938197
20487  1106  0.392992 -0.868924
24426   863 -0.199523 -0.831697
33665  1124  0.915199 -0.033524
34860  1122  0.305254  0.244650
37147  1122  0.982420 -0.252851
53573  1119 -0.457173 -0.575891
63077   874  0.914197 -0.133111
72263  1106  0.924446  0.187110
75129  1126 -0.820924 -0.831940
77157   874  0.286803 -0.908891
80028  1117  0.182344 -0.297617
85099  1124  0.454558  0.011887
93097   889  0.333730 -0.539914
93482  1106 -0.448811 -0.646975
```

Some things to keep in mind:

* The `groupby_filter` function above assumes the default index (`RangeIndex`) on the input dataframe `df` -- or at least, a monotonically increasing index.
* The groupby keys are always assumed to be column names of the input DataFrame
* In general, the performance will be good only for a small number of large groups, and quite bad for a large number of small groups.
